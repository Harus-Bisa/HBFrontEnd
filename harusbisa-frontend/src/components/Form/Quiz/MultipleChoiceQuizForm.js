import React from "react"
import { Form, FormGroup, Label, Input, Button as RButton, Col} from 'reactstrap';
import {connect} from "react-redux";
import { Button } from "@material-ui/core";
import AnswerForm from "./AnswerForm";
import { addQuiz, removeAnwers, setAnswer, setCorrectAnswer } from "../../../redux/actions";

function MultipleChoiceQuizForm(props){
    var [question, setQuestion] = React.useState(props.quiz ? props.quiz.question : "")
    var [numOfAnswers, setNumOfAnswers] = React.useState(props.quiz ? props.quiz.answerOptions.length : 2)
    var [points, setPoints] = React.useState(props.quiz ? props.quiz.pointWorth : 1)
    var [duration, setDuration] = React.useState(props.quiz ? props.quiz.duration : 60)
    const type = (props.id !== null && props.id !== undefined ? "EDIT" : "ADD")

    const submit = (event) =>{
        event.preventDefault();
        props.addQuiz(props.lectureId, question, props.answers,props.correctAnswer, duration, points)
    }
    const incrementNumOfAns = () =>{
        setNumOfAnswers(numOfAnswers + 1)
    }
    const makeAnswers = () =>{
        let answers = []
        for (let i=0; i<numOfAnswers; i++){
            answers.push(
                <AnswerForm option={i} key={i} numOfAnswers={numOfAnswers} setNumOfAnswers={setNumOfAnswers}/>
            )
        }
        return answers
    }

    React.useEffect(() => {
        if(type === "EDIT" && props.quiz){
            for(let i=0; i<props.quiz.answerOptions.length; i++){
                props.setAnswer(i, props.quiz.answerOptions[i])
            }
            props.setCorrectAnswer(props.quiz.correctAnswerIndex)
        }
        return () => {
            props.removeAnwers();
        }
      }, []);
    var verified = question !=="" && points !=="" && duration!=="" && props.answers!==[] && props.correctAnswer!==null
    return(
        <div className="container">
            <div className="row">
                <div className="col-12" style={{display:'flex', justifyContent:'center', padding:'1rem 2rem'}}>
                    <h3>{type === "ADD" ? "Tambah Pertanyaan": "Edit Pertanyaan"}</h3>
                </div>
            </div>
            <div className="content">
                <Form onSubmit={submit} style={{minWidth:'50vw'}}>
                    <FormGroup>
                        <Label>Pertanyaan*</Label>
                        <Input type="textarea" id="question" required value={question} placeholder={"Tulis pertanyaan Anda disini"} onChange={(event) => setQuestion(event.target.value)}/>
                    </FormGroup>
                    <FormGroup>
                        <Label>Jawaban*</Label>
                        {makeAnswers()}
                        <div style={{display:'flex', justifyContent:"flex-end"}}>
                            <RButton className="neutral-button" onClick={incrementNumOfAns}>Tambah Pilihan Jawaban</RButton>
                        </div>
                    </FormGroup>
                    <FormGroup>
                        <Label>Setting Pilihan Ganda</Label>
                        <div className="row justify-content-end">
                            <div className="col-11">
                                <FormGroup row>
                                    <Label id="point" sm={3}>Poin Soal Ini*</Label>
                                    <Col sm={9}>
                                        <Input
                                            type="number"
                                            value={points}
                                            onChange={(event) => setPoints(event.target.value)}
                                            required
                                        />
                                    </Col>
                                </FormGroup>
                                <FormGroup row>
                                    <Label sm={3}>Timer*</Label>
                                    <Col sm={6}>
                                        <Input
                                            type="number"
                                            value={duration}
                                            onChange={(event) => setDuration(event.target.value)}
                                            required
                                        />
                                    </Col>
                                    <Label sm={3}>Detik</Label>
                                </FormGroup>
                            </div>
                        </div>
                    </FormGroup>
                    
                    <div className="row justify-content-end">
                        {type === "ADD" && <div className="col-3">
                            <Button fullWidth className="prof-button" onClick={() => props.changeQuizType(null)}>Kembali</Button>
                        </div>}
                        <div className="col-3">
                            <Button fullWidth className="prof-button" type="submit" disabled={!verified}>{type === "ADD" ? "Tambahkan" : "Edit"}</Button>
                        </div>
                    </div>
                </Form>
            </div>
        </div>
    )
}

function mapStateToProps(state, ownProps){
    var targetQuiz = null;
    if (ownProps.id !== null && ownProps.id !== undefined){
        targetQuiz = state.selectedLecture.quizzes[ownProps.id]
    }
    return{
        lectureId: state.selectedLecture.lectureId,
        answers: state.answers,
        correctAnswer: state.correctAnswer,
        quiz: targetQuiz
    }
}
export default connect(mapStateToProps,{addQuiz, removeAnwers, setAnswer,setCorrectAnswer})(MultipleChoiceQuizForm);