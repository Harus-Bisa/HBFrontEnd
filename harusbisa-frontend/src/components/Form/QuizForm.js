import React from "react"
import { Form, FormGroup, Label, Input, Button as RButton, Col} from 'reactstrap';
import {connect} from "react-redux";
import { Button } from "@material-ui/core";
import AnswerForm from "./AnswerForm";

function QuizForm(props){
    var [question, setQuestion] = React.useState("")
    var [numOfAnswers, setNumOfAnswers] = React.useState(2)
    var [points, setPoints] = React.useState(1)
    var [time, setTime] = React.useState(60)
    
    const submit = (event) =>{
        event.preventDefault();
    }
    const incrementNumOfAns = () =>{
        setNumOfAnswers(numOfAnswers + 1)
    }
    const makeAnswers = () =>{
        let answers = []
        for (let i=0; i<numOfAnswers; i++){
            answers.push(
                <AnswerForm option={i} key={i}/>
            )
        }

        return answers
    }
    var verified = question && points
    return(
        <div className="container">
            <div className="row">
                <div className="col-12" style={{display:'flex', justifyContent:'center', padding:'1rem 2rem'}}>
                    <h3>Tambah Pertanyaan</h3>
                </div>
            </div>
            <div className="content">
                <Form onSubmit={submit} style={{minWidth:'50vw'}}>
                    <FormGroup>
                        <Label>Pertanyaan*</Label>
                        <Input type="textarea" id="question" required value={question} placeholder={"Tulis pertanyaan Anda disini"} onChange={(event) => setQuestion(event.target.value)}/>
                    </FormGroup>
                    <FormGroup>
                        <Label>Jawaban*</Label>
                        {makeAnswers()}
                        <div style={{display:'flex', justifyContent:"flex-end"}}>
                            <RButton className="neutral-button" onClick={incrementNumOfAns}>Tambah Pilihan Jawaban</RButton>
                        </div>
                    </FormGroup>
                    <FormGroup>
                        <Label>Setting Pilihan Ganda</Label>
                        <div className="row justify-content-end">
                            <div className="col-11">
                                <FormGroup row>
                                    <Label id="point" sm={3}>Poin Soal Ini*</Label>
                                    <Col sm={9}>
                                        <Input
                                            type="number"
                                            value={points}
                                            onChange={(event) => setPoints(event.target.value)}
                                            required
                                        />
                                    </Col>
                                </FormGroup>
                                <FormGroup row>
                                    <Label sm={3}>Timer*</Label>
                                    <Col sm={6}>
                                        <Input
                                            type="number"
                                            value={time}
                                            onChange={(event) => setTime(event.target.value)}
                                            required
                                        />
                                    </Col>
                                    <Label sm={3}>Detik</Label>
                                </FormGroup>
                            </div>
                        </div>
                    </FormGroup>
                    
                    <div className="row justify-content-end">
                        <div className="col-3">
                            <Button fullWidth className="prof-button">Kembali</Button>
                        </div>
                        <div className="col-3">
                            <Button fullWidth className="prof-button" type="submit" disabled={!verified}>Tambahkan</Button>
                        </div>
                    </div>
                </Form>
            </div>
        </div>
    )
}

export default connect()(QuizForm);