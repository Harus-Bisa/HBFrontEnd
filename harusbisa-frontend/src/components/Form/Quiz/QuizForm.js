import React from "react";
import QuizTypeForm from "./QuizTypeForm";
import MultipleChoiceQuizForm from "./MultipleChoiceQuizForm";

export const MULTIPLE_CHOICE = "MULTIPLE_CHOICE";

function QuizForm(props){
    const [quizType, setQuizType] = React.useState(null);
    const handleChange = (type) => {
        setQuizType(type)
    }
    return(
        <div>
            {!quizType && <QuizTypeForm changeQuizType={handleChange}/>}
            {quizType === MULTIPLE_CHOICE && <MultipleChoiceQuizForm changeQuizType={handleChange}/>}
        </div>
    )
}

export default QuizForm;