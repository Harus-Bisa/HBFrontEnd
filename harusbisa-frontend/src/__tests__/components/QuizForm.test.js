import React from "react";
import {mount} from "enzyme";
import configureMockStore from "redux-mock-store";
import '../../test-config';
import {Provider} from 'react-redux';
import thunk from 'redux-thunk';
import QuizForm from "../../components/Form/Quiz/QuizForm";
const mockStore = configureMockStore([thunk]);
import multipleChoicePic from "../../img/multipleChoice.png";
import wordPic from "../../img/word.png";
import MultipleChoiceQuizForm from "../../components/Form/Quiz/MultipleChoiceQuizForm";

describe("QuizForm component", () =>{
    let component;
    let store;
    let initialState;
    let mockClosePopup = jest.fn()
    beforeEach(() =>{
        initialState={
            selectedLecture:{
                lectureId:"1"
            }
        }
        store = mockStore(initialState);
        component = mount(<Provider store={store}><QuizForm/></Provider>)
    })
    afterEach(() =>{
        jest.clearAllMocks();
    })
    it("matches snapshot", () =>{
        expect(component.html()).toMatchSnapshot()
    })
    it("changes picture on hover", () => {
        let image = component.find("img")
        expect(image.prop("src")).toEqual(multipleChoicePic)

        let multipleChoiceButton = component.find("#multiple-choice").at(0)
        let wordButton = component.find("#word").at(0)
        let numericButton = component.find("#numeric").at(0)
        wordButton.simulate("mouseEnter")
        expect(component.find("img").prop("src")).toEqual(wordPic)
        multipleChoiceButton.simulate("mouseEnter")
        expect(component.find("img").prop("src")).toEqual(multipleChoicePic)
        numericButton.simulate("mouseEnter")
        expect(component.find("img").prop("src")).toEqual(multipleChoicePic)
    })
    it("renders correct type of form", () =>{
        expect(component.find(MultipleChoiceQuizForm)).toHaveLength(0)
        let multipleChoiceButton = component.find("#multiple-choice").at(0)
        multipleChoiceButton.simulate("click")
        expect(component.find(MultipleChoiceQuizForm)).toHaveLength(1)
    })
})