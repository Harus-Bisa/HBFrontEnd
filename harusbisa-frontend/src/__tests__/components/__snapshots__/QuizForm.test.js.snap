// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizForm component matches snapshot 1`] = `"<div><div class=\\"container\\"><div class=\\"row\\"><div class=\\"col-5 content\\"><div><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\" id=\\"multiple-choice\\"><span class=\\"MuiButton-label\\"><div style=\\"display: flex; flex-direction: column; text-align: start; margin-top: 15px; margin-bottom: 15px;\\"><h5>Pilihan Ganda</h5><p>Tanggal ini akan menjadi nama sesi anda</p></div></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\" id=\\"word\\"><span class=\\"MuiButton-label\\"><div style=\\"display: flex; flex-direction: column; text-align: start; margin-top: 15px; margin-bottom: 15px;\\"><h5>Isian</h5><p>Tanggal ini akan menjadi nama sesi anda</p></div></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\"><div style=\\"display: flex; flex-direction: column; text-align: start; margin-top: 15px; margin-bottom: 15px;\\" id=\\"numeric\\"><h5>Angka</h5><p>Tanggal ini akan menjadi nama sesi anda</p></div></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div><div class=\\"col-7\\" style=\\"background: rgb(239, 241, 245);\\"><div style=\\"display: flex; height: 100%; width: 100%;\\"><img src=\\"multipleChoice.png\\" alt=\\"Pilihan Ganda\\" style=\\"margin: auto;\\"></div></div></div></div></div>"`;
